AndroidXCI/lib/src/test/kotlin/dev/androidx/ci/testRunner/FirebaseTestLabControllerTest.kt

```diff
@@ -99,6 +100,16 @@ internal class FirebaseTestLabControllerTest {
         }
     }
 
+    @Test
```

Comment:
i don't think this test is adding any value. it just makes sure API is there and getting what the fake API will return.

rather, we need a proper test for `AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestExecutionStore.kt` that fills in the fake service (like we do for histories) and makes sure it does pull them properly (e.g. inclluding paging them all).

Reply:
Can you point to the tests for histories? I see some tests in "ToolsResultApiPlaygroundTest", but those are not based on fake results if I understand correctly.

Reply:
@yigit I have added a method to mock some data around steps and and a test to verify it. And also another test in `AndroidXCI/lib/src/test/kotlin/dev/androidx/ci/firebase/ToolsResultApiPlaygroundTest.kt`, I wasn't able to run it through the playground, only verified it though FTL apis directly, but pretty sure it'll work correctly. If you can point me to the "histories" test, I can add something similar for test execution steps too.

Reply:
Playground tests are not real tests. They are just convenient ways to call the real API to see the response types etc ( to be used during development or troubleshooting). It is also why they are ignored on CI.

As for the actual testing, there are various Fake implementations you can check. The history example was from `FakeToolsResultApi`. Or you can check tests using `FakeFirebaseTestLabApi`. 

For example, `FakeFirebaseTestLabApi` provides `setTestMatrix` which allows the test the update the "backend" TestMatrix value then invoke other methods that should ideally use that new value.

So in this case, i was thinking you expand the FakeToolsResultApi to provide some API to add steps and also be able to return them. Then you can do an end to end test which will need to query these steps and return them (to validate the result)

Another example is `TestRunnerTest`. For instance, `singleTest` creates an archive file on the fake backend, then invokes `testRunner.runTests()` which should be getting those fake zip file and creating tests out of it. Then the test validates the results from the fake backend.

So tl;dr; what I'm looking for is a test that prepares the backend state (add steps etc), run the top level API we have that needs to access to those steps and ensure that returned values include those steps. Which will also allow you to test cases where a result doesn't exist etc.

Reply:
@yigit Thank you for these references! I have updated `FakeToolsResultApi` to have a way for adding executionSteps for our tests. I also updated tests in `TestRunnerServiceImplTest` that run more end to end scenarios, with mock data that is inline with what FTL provides through its apis. It covers a positive case where testSteps are present in the api output and logcats are present in the glcoud bucket, as well as negative cases where either step is missing or actual logcat file is missing from gcloud bucket. I have also updated the test with shards and reruns and made sure that scenario is also working as expected.